lib/Sema/CSSimplify.cpp

```diff
@@ -2521,6 +2521,20 @@ ConstraintSystem::matchTypesBindTypeVar(
     }
   }
 
+  // If we are attempting to bind a function type
+  // to a keypath (to allow KeyPath<R,V> to be used in places where
+  // ((R)->V) is expected), it does not make sense to bind functions
+  // with more than 1 parameter.  
+  if (auto *loc = typeVar->getImpl().getLocator()) {
+    if (loc->isLastElement<LocatorPathElt::KeyPathType>()) {
+      if (auto *fnType = type->getAs<AnyFunctionType>()) {
+        if (fnType->getParams().size() != 1) {
+          return formUnsolvedResult();
```

Comment:
I think this a too late for check like this, we should keep this local to `simplifyKeyPathConstraint`. It's okay to let `simplifyKeyPathConstraint` fail in regular mode but in `shouldAttemptFixes` mode I think it should _pretend_ that key path got resolved and move on. One way to do that might be to record a `IgnoreContextualType` fix for this mismatch and pretend that there is no `keyPathTy` available (which also means that `definitelyFunctionType` could only be set after that check).